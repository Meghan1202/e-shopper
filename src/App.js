import React, { useEffect, useState } from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import Axios from 'axios';
import NavBar from './components/NavBar/NavBar';
import Home from './components/Home/Home';
import Cart from './components/Cart/Cart';
import Checkout from './components/Checkout/Checkout';
import Order from './components/Orders/Orders';
import ThemeContext, { Theme } from './ThemeContext';

const getAllInventory = async (url) => {
  const apiResponse = await Axios.get(url);
  const jsonResponse = apiResponse.data;
  const items = jsonResponse.data.map((item) => {
    const modifiedItem = item;
    modifiedItem.stock = item.count;
    modifiedItem.count = 0;
    modifiedItem.companyName = 'Bigbasket';
    modifiedItem.imgSrc = item.img ? item.img : 'https://complianz.io/wp-content/uploads/2019/03/placeholder-300x202.jpg';
    return item;
  });
  return items;
};

const App = () => {
  const [products, setProducts] = useState([]);

  const [cartItems, setCartItems] = useState([]);

  const [cartCount, setCartCount] = useState(0);

  useEffect(async () => {
    const inventory = await getAllInventory('/items');
    setProducts(inventory);
  }, []);

  const onIncrement = (product) => {
    setCartCount(
      product.stock > 0 ? cartCount + 1 : cartCount,
    );
    setProducts(
      products.map((item) => {
        const modifiedItem = item;
        if (item.id === product.id && item.stock > 0) {
          modifiedItem.stock -= 1;
          modifiedItem.count += 1;
        }
        return modifiedItem;
      }),
    );
  };

  const onDecrement = (product) => {
    setCartCount(
      product.count > 0 ? cartCount + 1 : cartCount,
    );
    setProducts(
      products.map((item) => {
        const modifiedItem = item;
        if (item.id === product.id && item.count > 0) {
          modifiedItem.stock += 1;
          modifiedItem.count -= 1;
        }
        return modifiedItem;
      }),
    );
  };

  const [theme, setTheme] = useState('white');

  const handleThemeChange = () => {
    if (theme === 'white') {
      setTheme('black');
    } else {
      setTheme('white');
    }
  };

  return (
    <>
      <input type="checkbox" value="Dark mode" onClick={handleThemeChange} />
      <span>{theme === 'white' ? 'Light mode' : 'Dark mode'}</span>
      <BrowserRouter>
        <ThemeContext.Provider value={theme === 'white' ? Theme.light : Theme.dark}>
          <NavBar items={cartCount} />
        </ThemeContext.Provider>
        <Switch>
          <Route path="/" exact>
            <Home
              products={products}
              onDecrement={onDecrement}
              onIncrement={onIncrement}
            />
          </Route>
          <Route path="/cart"><Cart cartItems={cartItems} /></Route>
          <Route path="/checkout"><Checkout /></Route>
          <Route path="/allOrder"><Order noOfItems={cartItems.length} cartItems={cartItems} /></Route>
        </Switch>
      </BrowserRouter>
    </>
  );
};

export default App;
