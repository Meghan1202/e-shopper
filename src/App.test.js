import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import App from './App';

describe(App.name, () => {
  test('should match the snapshot', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
  test('should render the NavBar', () => {
    render(<App />);
    const clickOnNavBar = screen.getByTestId('nav-bar');
    fireEvent.click(clickOnNavBar);
    expect(document.location.href).toBe('http://localhost/');
  });
  test('should render the home component and route to home page', () => {
    render(<App />);
    screen.getByTestId('home-page');
  });
});
