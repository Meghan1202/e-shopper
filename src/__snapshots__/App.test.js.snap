// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div>
  <input
    type="checkbox"
    value="Dark mode"
  />
  <span>
    Light mode
  </span>
  <div
    class="topnav__light"
    data-testid="nav-bar"
  >
    <a
      class="app__logo"
      href="/"
    >
      E-Shopper
    </a>
    <br />
    <div>
      <a
        href="/allOrder"
      >
        All Orders
      </a>
      <a
        class="active"
        href="/cart"
      >
        <img
          alt="Basket"
          id="basket__image"
          src="https://img.freepik.com/free-vector/empty-wicker-basket-flowers-large-birds-nest-eggs_135176-431.jpg?size=626&ext=jpg"
        />
        My Basket
        <span>
          0
        </span>
      </a>
    </div>
  </div>
  <div
    data-testid="home-page"
  >
    <h3>
       Fruits And Vegitables
    </h3>
    <div
      class="card__row"
    >
      <div
        class="card"
        data-testid="product-card"
      >
        <img
          alt="Banana"
          data-testid="product-image"
          id="banana__img"
          src="assets/fruits/banana.png"
        />
        <div
          class="container"
        >
          <h4>
            Fresho
          </h4>
          <p>
            <b>
              Banana
            </b>
          </p>
          <p>
            Quantity:
            1 kg
          </p>
          <div
            class="inner__card"
          >
            <p>
              40
              /-
            </p>
            <div
              class="quantity__count"
              data-testid="quantity-count"
            >
              <button
                type="button"
              >
                -
              </button>
              <p>
                0 in Basket
              </p>
              <button
                type="button"
              >
                 + 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card"
        data-testid="product-card"
      >
        <img
          alt="Banana"
          data-testid="product-image"
          id="banana__img"
          src="assets/fruits/mango.png"
        />
        <div
          class="container"
        >
          <h4>
            Fresho
          </h4>
          <p>
            <b>
              Mango
            </b>
          </p>
          <p>
            Quantity:
            1 kg
          </p>
          <div
            class="inner__card"
          >
            <p>
              40
              /-
            </p>
            <div
              class="quantity__count"
              data-testid="quantity-count"
            >
              <button
                type="button"
              >
                -
              </button>
              <p>
                0 in Basket
              </p>
              <button
                type="button"
              >
                 + 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card"
        data-testid="product-card"
      >
        <img
          alt="Banana"
          data-testid="product-image"
          id="banana__img"
          src="assets/fruits/orange.png"
        />
        <div
          class="container"
        >
          <h4>
            Fresho
          </h4>
          <p>
            <b>
              Orange
            </b>
          </p>
          <p>
            Quantity:
            1 kg
          </p>
          <div
            class="inner__card"
          >
            <p>
              40
              /-
            </p>
            <div
              class="quantity__count"
              data-testid="quantity-count"
            >
              <button
                type="button"
              >
                -
              </button>
              <p>
                0 in Basket
              </p>
              <button
                type="button"
              >
                 + 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card"
        data-testid="product-card"
      >
        <img
          alt="Banana"
          data-testid="product-image"
          id="banana__img"
          src="assets/fruits/apple.png"
        />
        <div
          class="container"
        >
          <h4>
            Fresho
          </h4>
          <p>
            <b>
              Apple
            </b>
          </p>
          <p>
            Quantity:
            1 kg
          </p>
          <div
            class="inner__card"
          >
            <p>
              40
              /-
            </p>
            <div
              class="quantity__count"
              data-testid="quantity-count"
            >
              <button
                type="button"
              >
                -
              </button>
              <p>
                0 in Basket
              </p>
              <button
                type="button"
              >
                 + 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card"
        data-testid="product-card"
      >
        <img
          alt="Banana"
          data-testid="product-image"
          id="banana__img"
          src="assets/fruits/strawberry.png"
        />
        <div
          class="container"
        >
          <h4>
            Fresho
          </h4>
          <p>
            <b>
              Strawberry
            </b>
          </p>
          <p>
            Quantity:
            1 kg
          </p>
          <div
            class="inner__card"
          >
            <p>
              40
              /-
            </p>
            <div
              class="quantity__count"
              data-testid="quantity-count"
            >
              <button
                type="button"
              >
                -
              </button>
              <p>
                0 in Basket
              </p>
              <button
                type="button"
              >
                 + 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card"
        data-testid="product-card"
      >
        <img
          alt="Banana"
          data-testid="product-image"
          id="banana__img"
          src="assets/fruits/strawberry.png"
        />
        <div
          class="container"
        >
          <h4>
            Fresho
          </h4>
          <p>
            <b>
              Strawberry
            </b>
          </p>
          <p>
            Quantity:
            1 kg
          </p>
          <div
            class="inner__card"
          >
            <p>
              40
              /-
            </p>
            <div
              class="quantity__count"
              data-testid="quantity-count"
            >
              <button
                type="button"
              >
                -
              </button>
              <p>
                0 in Basket
              </p>
              <button
                type="button"
              >
                 + 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card"
        data-testid="product-card"
      >
        <img
          alt="Banana"
          data-testid="product-image"
          id="banana__img"
          src="assets/fruits/strawberry.png"
        />
        <div
          class="container"
        >
          <h4>
            Fresho
          </h4>
          <p>
            <b>
              Strawberry
            </b>
          </p>
          <p>
            Quantity:
            1 kg
          </p>
          <div
            class="inner__card"
          >
            <p>
              40
              /-
            </p>
            <div
              class="quantity__count"
              data-testid="quantity-count"
            >
              <button
                type="button"
              >
                -
              </button>
              <p>
                0 in Basket
              </p>
              <button
                type="button"
              >
                 + 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card"
        data-testid="product-card"
      >
        <img
          alt="Banana"
          data-testid="product-image"
          id="banana__img"
          src="assets/fruits/strawberry.png"
        />
        <div
          class="container"
        >
          <h4>
            Fresho
          </h4>
          <p>
            <b>
              Strawberry
            </b>
          </p>
          <p>
            Quantity:
            1 kg
          </p>
          <div
            class="inner__card"
          >
            <p>
              40
              /-
            </p>
            <div
              class="quantity__count"
              data-testid="quantity-count"
            >
              <button
                type="button"
              >
                -
              </button>
              <p>
                0 in Basket
              </p>
              <button
                type="button"
              >
                 + 
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
